File: n_puzle/busqueda.c
/*******************************************/
/* 		    BUSQUEDA.C                     */
/*						                   */
/* Asignatura: Inteligencia Artificial     */
/* Grado en Ingenieria Informatica - UCA   */
/*******************************************/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "puzle.h"
#include "listaia.h"
#include "busqueda.h"

void solucionFin(int res){
   printf("\nFin de la busqueda\n");
    if (res)
      printf("Se ha llegado al objetivo\n");
   else
      printf("No se ha llegado al objetivo\n");
}
void dispNodo(tNodo *nodo){
    dispEstado(nodo->estado);
    printf("\n");
}
void dispCamino(tNodo *nodo){
    if (nodo->padre==NULL){
        printf("\n Desde el inicio\n");
        dispEstado(nodo->estado);
    }
    else {
        dispCamino(nodo->padre);
         dispOperador(nodo->operador);
        dispEstado(nodo->estado);
    }
}

void dispSolucion(tNodo *nodo){
   printf("Profundidad=%d\n",nodo->profundidad);
   printf("Coste=%d\n",nodo->costeCamino);
   dispCamino(nodo);
}


/* Crea el nodo raíz. */
tNodo *nodoInicial(){
   tNodo *inicial=(tNodo *) malloc(sizeof(tNodo));
   inicial->estado=estadoInicial();
   inicial->padre=NULL;
   inicial->costeCamino=0;
   inicial->profundidad=0;
   return inicial;
}

LISTA expandir(tNodo *nodo, int fun){
	unsigned op;
	LISTA sucesores=VACIA;
	tNodo *nuevo=calloc(1,sizeof(tNodo));
	tEstado *s;
	for (op=1; op<=NUM_OPERADORES;op++){
		if (esValido(op,nodo->estado)){
			//s=(tEstado *)calloc(1,sizeof(tEstado));
			s=aplicaOperador(op,nodo->estado);
			nuevo->estado=s;
			nuevo->padre=nodo;
			nuevo->operador=op;
			nuevo->costeCamino=nodo->costeCamino + coste(op,nodo->estado);
			nuevo->profundidad=nodo->profundidad+1;
			switch(fun){
			case 1:
				nuevo->valHeuristica=heuristica(s,estadoObjetivo());
				break;
			case 2:
				nuevo->valHeuristica=heuristica(s,estadoObjetivo())+camino(nuevo);
				break;
			case 3:
				nuevo->valHeuristica=manhattan(s,estadoObjetivo());
				break;
			case 4:
				nuevo->valHeuristica=manhattan(s,estadoObjetivo())+camino(nuevo);
				break;
			default:
				nuevo->valHeuristica=0;
				break;
			}
			//printf("\nHeurística no lista= %d",nuevo->valHeuristica);
			InsertarUltimo(&sucesores,  (tNodo *) nuevo, (sizeof (tNodo)));
		}
	}
	return sucesores;
}

int heuristica(tEstado* e, tEstado* final){	// Número de piezas mal colocadas
	int h=0;

	for(int i=0; i<N*N; i++){
		if(final->col[i]!=e->col[i] || final->fila[i]!=e->fila[i]){
			h++;
		}
	}

	return h;
}

int camino(tNodo* n){	// Profundidad del nodo
	int g=0;
	tNodo* aux=n;

	while(aux!=NULL){
		aux=aux->padre;
		g++;
	}

	return g;
}

int manhattan(tEstado* e, tEstado* final){	// Distancia de Manhattan
	int d=0;

	for(int i=0; i<N*N; i++){
		d+=abs(e->col[i]-final->col[i])+abs(e->fila[i]-final->fila[i]);
	}

	return d;
}

//###################################################################################################
int busquedaAnch(){
    int objetivo=0, visitados=0;
    tNodo *Actual=(tNodo*) calloc(1,sizeof(tNodo));
    tNodo *Inicial=nodoInicial();

    LISTA Abiertos= VACIA;
    LISTA Sucesores= VACIA;
    InsertarPrimero(&Abiertos,(tNodo*) Inicial,sizeof(tNodo));
    while (!esVacia(Abiertos) && !objetivo){
        Actual=(tNodo*) calloc(1,sizeof(tNodo));
        ExtraerPrimero(Abiertos,Actual, sizeof(tNodo));
        EliminarPrimero(&Abiertos);
        visitados++;
        objetivo=testObjetivo(Actual->estado);
        if (!objetivo){
            Sucesores = expandir(Actual,0);
            Abiertos=Concatenar(Abiertos,Sucesores);
      }
   }//while
   
    printf("\nVisitados= %d\n", visitados);
    if (objetivo)
        dispSolucion(Actual);
    /*while(esVacia(Sucesores)){
      	EliminarPrimero(Sucesores);
    }*/
    //free(Sucesores);
    free(Inicial);
    free(Actual);
    return objetivo;
}

//###################################################################################################
int busquedaProf(){
    int objetivo=0, visitados=0;
    tNodo *Actual=(tNodo*) calloc(1,sizeof(tNodo));
    tNodo *Inicial=nodoInicial();

    LISTA Abiertos= VACIA;
    LISTA Sucesores= VACIA;
    InsertarPrimero(&Abiertos,(tNodo*) Inicial,sizeof(tNodo));
    while (!esVacia(Abiertos) && !objetivo){
        Actual=(tNodo*) calloc(1,sizeof(tNodo));
        ExtraerPrimero(Abiertos,Actual, sizeof(tNodo));
        EliminarPrimero(&Abiertos);
        visitados++;
        objetivo=testObjetivo(Actual->estado);
        if (!objetivo){
            Sucesores = expandir(Actual,0);
            Abiertos=Concatenar(Sucesores,Abiertos);
      }
   }//while

    printf("\nVisitados= %d\n", visitados);
    if (objetivo)
        dispSolucion(Actual);
    /*while(esVacia(Sucesores)){
    	EliminarPrimero(Sucesores);
    }*/
    //free(Sucesores);
    free(Inicial);
    free(Actual);
    return objetivo;
}

//###################################################################################################
int busquedaAnchEstRep(){
    int objetivo=0, visitados=0;
    tNodo *Actual=(tNodo*) calloc(1,sizeof(tNodo));
    tNodo *Inicial=nodoInicial();

    LISTA Abiertos= VACIA;
    LISTA Sucesores= VACIA;
    LISTA Cerrados= VACIA;
    InsertarPrimero(&Abiertos,(tNodo*) Inicial,sizeof(tNodo));
    while (!esVacia(Abiertos) && !objetivo){
        Actual=(tNodo*) calloc(1,sizeof(tNodo));
        ExtraerPrimero(Abiertos,Actual, sizeof(tNodo));
        EliminarPrimero(&Abiertos);
        if(!buscarElto(&Cerrados, Actual->estado)){
        	visitados++;
        	objetivo=testObjetivo(Actual->estado);
        	if (!objetivo){
        	    Sucesores = expandir(Actual,0);
        	    Abiertos=Concatenar(Abiertos,Sucesores);
        	}
        	InsertarUltimo(&Cerrados,(tNodo*) Actual, sizeof(tNodo));
        }
   }//while

    printf("\nVisitados= %d\n", visitados);
    if (objetivo)
        dispSolucion(Actual);
    /*while(esVacia(Sucesores)){
      	EliminarPrimero(Sucesores);
    }
    while(esVacia(Cerrados)){
      	EliminarPrimero(Cerrados);
    }*/
    //free(Sucesores);
    free(Inicial);
    free(Actual);
    return objetivo;
}

//###################################################################################################
int busquedaProfEstRep(){
    int objetivo=0, visitados=0;
    tNodo *Actual=(tNodo*) calloc(1,sizeof(tNodo));
    tNodo *Inicial=nodoInicial();

    LISTA Abiertos= VACIA;
    LISTA Sucesores= VACIA;
    LISTA Cerrados= VACIA;
    InsertarPrimero(&Abiertos,(tNodo*) Inicial,sizeof(tNodo));
    while (!esVacia(Abiertos) && !objetivo){
        Actual=(tNodo*) calloc(1,sizeof(tNodo));
        ExtraerPrimero(Abiertos,Actual, sizeof(tNodo));
        EliminarPrimero(&Abiertos);
        if(!buscarElto(&Cerrados, Actual->estado)){
        	visitados++;
        	objetivo=testObjetivo(Actual->estado);
        	if (!objetivo){
        	    Sucesores = expandir(Actual,0);
        	    Abiertos=Concatenar(Sucesores,Abiertos);
        	}
        	InsertarPrimero(&Cerrados,(tNodo*) Actual, sizeof(tNodo));
        }
   }//while

    printf("\nVisitados= %d\n", visitados);
    if (objetivo)
        dispSolucion(Actual);
    /*while(esVacia(Sucesores)){
      	EliminarPrimero(Sucesores);
    }
    while(esVacia(Cerrados)){
      	EliminarPrimero(Cerrados);
    }*/
    //free(Sucesores);
    free(Inicial);
    free(Actual);
    return objetivo;
}

//###################################################################################################
int busquedaAnchLimite(int l){
    int objetivo=0, visitados=0;
    tNodo *Actual=(tNodo*) calloc(1,sizeof(tNodo));
    tNodo *Inicial=nodoInicial();

    LISTA Abiertos= VACIA;
    LISTA Sucesores= VACIA;
    InsertarPrimero(&Abiertos,(tNodo*) Inicial,sizeof(tNodo));
    while (!esVacia(Abiertos) && !objetivo){
        Actual=(tNodo*) calloc(1,sizeof(tNodo));
        ExtraerPrimero(Abiertos,Actual, sizeof(tNodo));
        EliminarPrimero(&Abiertos);
        if(Actual->profundidad <= l){
        	visitados++;
        	objetivo=testObjetivo(Actual->estado);
        	if (!objetivo){
        	    Sucesores = expandir(Actual,0);
        	    Abiertos=Concatenar(Abiertos,Sucesores);
        	}
        }
   }//while

    printf("\nVisitados= %d\n", visitados);
    if (objetivo)
        dispSolucion(Actual);
    /*while(esVacia(Sucesores)){
      	EliminarPrimero(Sucesores);
    }*/
    //free(Sucesores);
    free(Inicial);
    free(Actual);
    return objetivo;
}

//###################################################################################################
int busquedaProfLimite(int l){
    int objetivo=0, visitados=0;
    tNodo *Actual=(tNodo*) calloc(1,sizeof(tNodo));
    tNodo *Inicial=nodoInicial();

    LISTA Abiertos= VACIA;
    LISTA Sucesores= VACIA;
    InsertarPrimero(&Abiertos,(tNodo*) Inicial,sizeof(tNodo));
    while (!esVacia(Abiertos) && !objetivo){
        Actual=(tNodo*) calloc(1,sizeof(tNodo));
        ExtraerPrimero(Abiertos,Actual, sizeof(tNodo));
        EliminarPrimero(&Abiertos);
        if(Actual->profundidad <= l){
        	visitados++;
        	objetivo=testObjetivo(Actual->estado);
        	if (!objetivo){
        	    Sucesores = expandir(Actual,0);
        	    Abiertos=Concatenar(Sucesores,Abiertos);
        	}
        }
   }//while

    printf("\nVisitados= %d\n", visitados);
    if (objetivo)
        dispSolucion(Actual);
    /*while(esVacia(Sucesores)){
      	EliminarPrimero(Sucesores);
    }*/
    //free(Sucesores);
    free(Inicial);
    free(Actual);
    return objetivo;
}

//###################################################################################################
int busquedaHeuristica(int op){
	int objetivo=0, visitados=0, max_abiertos=0, max_aux=0, generados=0;
	tNodo *Actual=(tNodo*) calloc(1,sizeof(tNodo));
	tNodo *Inicial=nodoInicial();

	LISTA Abiertos= VACIA;
	LISTA Sucesores= VACIA;
	LISTA Cerrados= VACIA;
	InsertarPrimero(&Abiertos,(tNodo*) Inicial,sizeof(tNodo));
	while (!esVacia(Abiertos) && !objetivo){
		Actual=(tNodo*) calloc(1,sizeof(tNodo));
		ExtraerPrimero(Abiertos,Actual, sizeof(tNodo));
		EliminarPrimero(&Abiertos);
		objetivo=testObjetivo(Actual->estado);
		if(!objetivo){
			if(!buscarElto(&Cerrados, Actual->estado)){
				visitados++;
				Sucesores = expandir(Actual,op);
				generados+=Longitud(Sucesores);

				Abiertos=Ordenar(Abiertos,Sucesores);

				max_aux=Longitud(Abiertos);
				if(max_aux>max_abiertos){
					max_abiertos=max_aux;
				}
				//printLista(Abiertos);
			}
			InsertarPrimero(&Cerrados,(tNodo*) Actual, sizeof(tNodo));
		}
	}//while

	printf("\nVisitados= %d\nGenerados= %d\nMáximo de abiertos= %d\n", visitados,generados,max_abiertos);
	if (objetivo)
		dispSolucion(Actual);
	free(Inicial);
	free(Actual);
	return objetivo;
}

File: cocheIA/app/src/main/java/ia/jfsr00/cocheia/Busqueda.java
package ia.jfsr00.cocheia;

import java.util.LinkedList;

public class Busqueda {
	public static NodoBusqueda anchura(){
		int visitados = 0;
		boolean objetivo = false;
		NodoBusqueda Actual = new NodoBusqueda();
		LinkedList<NodoBusqueda> Abiertos = new LinkedList<>(), Sucesores;

		Abiertos.add(Actual);
		while(!Abiertos.isEmpty() && !objetivo){
			Actual = Abiertos.getFirst();
			Abiertos.pollFirst();
			visitados++;

			objetivo = Actual.estado.testObjetivo();
			if(!objetivo){
				Sucesores = expandir(Actual);
				Abiertos.addAll(Sucesores);
			}

		}

		System.out.println("visitados: " + visitados);

		if(objetivo){
			return Actual;
		}else{
			return null;
		}
	}

	public static NodoBusqueda anchuraLim(int l){
		int visitados = 0;
		boolean objetivo = false;
		NodoBusqueda Actual = new NodoBusqueda();
		LinkedList<NodoBusqueda> Abiertos = new LinkedList<>(), Sucesores;

		Abiertos.add(Actual);
		while(!Abiertos.isEmpty() && !objetivo){
			Actual = Abiertos.getFirst();
			Abiertos.pollFirst();
			if(Actual.profundidad <= l){
				visitados++;
				objetivo = Actual.estado.testObjetivo();
				if(!objetivo){
					Sucesores = expandir(Actual);
					Abiertos.addAll(Sucesores);
				}
			}
		}
		System.out.println("visitados: " + visitados);
		if(objetivo){
			return Actual;
		}else{
			return null;
		}
	}

	public static NodoBusqueda anchuraEstRep(){
		int visitados = 0;
		boolean objetivo = false;
		NodoBusqueda Actual = new NodoBusqueda();
		LinkedList<NodoBusqueda> Abiertos = new LinkedList<>(), Cerrados = new LinkedList<>(), Sucesores;

		Abiertos.add(Actual);
		while(!Abiertos.isEmpty() && !objetivo){
			Actual = Abiertos.getFirst();
			Abiertos.pollFirst();
			if(!existsElto(Cerrados, Actual)){
				visitados++;
				objetivo = Actual.estado.testObjetivo();
				if(!objetivo){
					Sucesores = expandir(Actual);
					Abiertos.addAll(Sucesores);
				}
				Cerrados.addLast(Actual);
			}
		}
		System.out.println("visitados: " + visitados);
		if(objetivo){
			return Actual;
		}else{
			return null;
		}
	}

	public static NodoBusqueda profundidad(){
		int visitados = 0;
		boolean objetivo = false;
		NodoBusqueda Actual = new NodoBusqueda();
		LinkedList<NodoBusqueda> Abiertos = new LinkedList<>(), Sucesores;

		Abiertos.add(Actual);
		while(!Abiertos.isEmpty() && !objetivo){
			Actual = Abiertos.getFirst();
			Abiertos.pollFirst();
			visitados++;

			objetivo = Actual.estado.testObjetivo();
			if(!objetivo){
				Sucesores = expandir(Actual);
				Abiertos.addAll(0, Sucesores);
			}

		}

		System.out.println("visitados: " + visitados);

		if(objetivo){
			return Actual;
		}else{
			return null;
		}
	}

	public static NodoBusqueda profundidadLim(int l){
		int visitados = 0;
		boolean objetivo = false;
		NodoBusqueda Actual = new NodoBusqueda();
		LinkedList<NodoBusqueda> Abiertos = new LinkedList<>(), Sucesores;

		Abiertos.add(Actual);
		while(!Abiertos.isEmpty() && !objetivo){
			Actual = Abiertos.getFirst();
			Abiertos.pollFirst();
			if(Actual.profundidad <= l){
				visitados++;
				objetivo = Actual.estado.testObjetivo();
				if(!objetivo){
					Sucesores = expandir(Actual);
					Abiertos.addAll(0, Sucesores);
				}
			}
		}
		System.out.println("visitados: " + visitados);
		if(objetivo){
			return Actual;
		}else{
			return null;
		}
	}

	public static NodoBusqueda profundidadEstRep(){
		int visitados = 0;
		boolean objetivo = false;
		NodoBusqueda Actual = new NodoBusqueda();
		LinkedList<NodoBusqueda> Abiertos = new LinkedList<>(), Cerrados = new LinkedList<>(), Sucesores;

		Abiertos.add(Actual);
		while(!Abiertos.isEmpty() && !objetivo){
			Actual = Abiertos.getFirst();
			Abiertos.pollFirst();
			if(!existsElto(Cerrados, Actual)){
				visitados++;
				objetivo = Actual.estado.testObjetivo();
				if(!objetivo){
					Sucesores = expandir(Actual);
					Abiertos.addAll(0, Sucesores);
				}
				Cerrados.addLast(Actual);
			}
		}
		System.out.println("visitados: " + visitados);
		if(objetivo){
			return Actual;
		}else{
			return null;
		}
	}

	public static NodoBusqueda heuristica(int fun){
		int visitados = 0, max_abiertos = 0, max_aux = 0, generados = 0;
		boolean objetivo = false;
		NodoBusqueda Actual = new NodoBusqueda();
		LinkedList<NodoBusqueda> Abiertos = new LinkedList<>(), Cerrados = new LinkedList<>(), Sucesores;

		Abiertos.add(Actual);
		while(!Abiertos.isEmpty() && !objetivo){
			Actual = Abiertos.getFirst();
			Abiertos.pollFirst();
			objetivo = Actual.estado.testObjetivo();
			if(!objetivo){
				if(!existsElto(Cerrados, Actual)){
					visitados++;
					Sucesores = expandir(Actual, fun);
					generados += Sucesores.size();

					Abiertos = ordenar(Abiertos, Sucesores);

					max_aux = Abiertos.size();
					if(max_aux > max_abiertos){
						max_abiertos = max_aux;
					}
					Cerrados.addLast(Actual);
				}
			}
		}
		System.out.println("Visitados: " + visitados);
		System.out.println("Generados: " + generados);
		System.out.println("Máximo de abiertos: " + max_abiertos);
		if(objetivo){
			return Actual;
		}else{
			return null;
		}
	}

	public static LinkedList<NodoBusqueda> expandir(NodoBusqueda nodo, int fun){
		LinkedList<NodoBusqueda> sucesores = new LinkedList<>();
		NodoBusqueda nuevo;

		for(int op = 0; op < Coche.N_OP; op++){
			if(nodo.estado.esValido(op)){
				nuevo = new NodoBusqueda(new Coche(nodo.estado.aplicaOperador(op).getAsientos()));
				nuevo.padre = nodo;
				nuevo.operador = op;
				nuevo.costeCamino = nodo.costeCamino + nodo.estado.coste(op);
				nuevo.profundidad = nodo.profundidad + 1;
				switch(fun){
					case 1:
						nuevo.valHeuristica = nuevo.estado.heuristica();
						break;
					case 2:
						nuevo.valHeuristica = nuevo.estado.heuristica() + nuevo.camino();
						break;
					default:
						nuevo.valHeuristica=0;
						break;
				}

				sucesores.addLast(nuevo);
			}
		}
		return sucesores;
	}

	public static LinkedList<NodoBusqueda> expandir(NodoBusqueda nodo){
		return expandir(nodo, 0);
	}

	public static String caminoToString(NodoBusqueda nodo){
		String str = "";

		if(nodo == null){
			str += "\nSIN SOLUCIÓN\n";
		}else if(nodo.padre == null){
			str += "\nDesde el inicio\n" + nodo.estado.toString();
		}else{
			str += caminoToString(nodo.padre) + "\n Operador: " + nodo.estado.operadorToString(nodo.operador) + "\n" + nodo.estado.toString();
		}

		return str;
	}

	public static boolean existsElto(LinkedList<NodoBusqueda> l, NodoBusqueda n){
		boolean res = false;

		if(!l.isEmpty()){
			for(NodoBusqueda elto: l){
				if(elto.estado.iguales(n.estado)){
					res = true;
					break;
				}
			}
		}

		return res;
	}

	public static LinkedList<NodoBusqueda> insertarOrdenado(LinkedList<NodoBusqueda> C, NodoBusqueda nuevo){
		LinkedList<NodoBusqueda> R = new LinkedList<>();
		NodoBusqueda nc;

		if(C.isEmpty()){
			R.addLast(nuevo);
		}else{
			nc = C.getFirst();
			while(!C.isEmpty() && nuevo.valHeuristica >= nc.valHeuristica){
				R.addLast(nc);
				C.pollFirst();
				if(!C.isEmpty()){
					nc = C.getFirst();
				}
			}
			R.addLast(nuevo);
			R.addAll(C);
		}
		return R;
	}

	public static LinkedList<NodoBusqueda> ordenar(LinkedList<NodoBusqueda> A, LinkedList<NodoBusqueda> B){
		while(!B.isEmpty()){
			A = insertarOrdenado(A, B.getFirst());
			B.pollFirst();
		}

		return A;
	}
}
